executor/src/lib.rs

```diff
@@ -752,72 +752,91 @@ where
+                        let txn_info = TransactionInfo::new(
+                            txn.hash(),
+                            state_tree.root_hash(),
+                            event_tree.root_hash(),
+                            vm_output.gas_used(),
+                            status.major_status,
+                        );
+
+                        let real_txn_info_hash = txn_info.hash();
+                        txn_info_hashes.push(real_txn_info_hash);
+                        txn_info_hash = Some(real_txn_info_hash);
+                    }
+                    TransactionStatus::Discard(status) => {
+                        if !vm_output.write_set().is_empty() || !vm_output.events().is_empty() {
+                            crit!(
+                                "Discarded transaction has non-empty write set or events. \
+                                 Transaction: {:?}. Status: {}.",
+                                txn,
+                                status,
+                            );
+                        }
                     }
+                    TransactionStatus::Retry => (),
                 }
-            }
 
-            txn_data.push(TransactionData::new(
-                blobs,
-                vm_output.events().to_vec(),
-                vm_output.status().clone(),
-                Arc::clone(&state_tree),
-                Arc::new(event_tree),
-                vm_output.gas_used(),
-                num_accounts_created,
-                txn_info_hash,
-            ));
-            current_state_tree = state_tree;
-
-            // check for change in validator set
-            let validator_set_change_event_key = ValidatorSet::change_event_key();
-            for event in vm_output.events() {
-                if *event.key() == validator_set_change_event_key {
-                    next_validator_set = Some(ValidatorSet::from_bytes(event.event_data())?);
-                    break;
+                txn_data.push(TransactionData::new(
+                    blobs,
+                    vm_output.events().to_vec(),
+                    vm_output.status().clone(),
+                    Arc::clone(&state_tree),
+                    Arc::new(event_tree),
+                    vm_output.gas_used(),
+                    num_accounts_created,
+                    txn_info_hash,
+                ));
+                current_state_tree = state_tree;
+
+                // check for change in validator set
+                let validator_set_change_event_key = ValidatorSet::change_event_key();
+                for event in vm_output.events() {
+                    if *event.key() == validator_set_change_event_key {
+                        next_validator_set = Some(ValidatorSet::from_bytes(event.event_data())?);
+                        break;
+                    }
+                }
+
+                if next_validator_set.is_some() {
```

Comment:
similar here, there's no need to introduce the validator_set_change_seen value, we could just use the next_validator_set.is_some above might improve readability

also I would consider validator_set_change_seen restart_due_to_reconfiguration, which seems to be a bit more future focused. cc: @zekun000

Reply:
I'll just use `reconfiguration_seen`